gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,366 @@
+   *
+   */
+  @Override
+  public Collection<ProducerJob> getAllPages(String startDate, String endDate, String country, int rowLimit)
+      throws IOException {
+    if (!_jobs.isEmpty()) {
+      LOG.info("Service got hot started.");
+      return _jobs;
+    }
+
+    ApiDimensionFilter countryFilter = GoogleWebmasterFilter.countryEqFilter(country);
+
+    List<GoogleWebmasterFilter.Dimension> requestedDimensions = new ArrayList<>();
+    requestedDimensions.add(GoogleWebmasterFilter.Dimension.PAGE);
+
+    Collection<String> allPages =
+        _client.getPages(_siteProperty, startDate, endDate, country, rowLimit, requestedDimensions,
+            Arrays.asList(countryFilter), 0);
+    int actualSize = allPages.size();
+
+    if (rowLimit < GoogleWebmasterClient.API_ROW_LIMIT || actualSize < GoogleWebmasterClient.API_ROW_LIMIT) {
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    } else {
+      int expectedSize = getPagesSize(startDate, endDate, country, requestedDimensions, Arrays.asList(countryFilter));
+      LOG.info(String.format("Total number of pages is %d for market-%s from %s to %s", expectedSize,
+          GoogleWebmasterFilter.countryFilterToString(countryFilter), startDate, endDate));
+      Queue<Pair<String, FilterOperator>> jobs = new ArrayDeque<>();
+      expandJobs(jobs, _siteProperty);
+
+      allPages = getPages(startDate, endDate, requestedDimensions, countryFilter, jobs);
+      allPages.add(_siteProperty);
+      actualSize = allPages.size();
+      if (actualSize != expectedSize) {
+        LOG.warn(
+            String.format("Expected page size for country-%s is %d, but only able to get %d", country, expectedSize,
+                actualSize));
+      }
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    }
+
+    ArrayDeque<ProducerJob> jobs = new ArrayDeque<>(actualSize);
+    for (String page : allPages) {
+      jobs.add(new SimpleProducerJob(page, startDate, endDate));
+    }
+    return jobs;
+  }
+
+  private int getPagesSize(final String startDate, final String endDate, final String country,
+      final List<Dimension> requestedDimensions, final List<ApiDimensionFilter> apiDimensionFilters)
+      throws IOException {
+
+    int startRow = 0;
+    int count = 4;
+    final ExecutorService es = Executors.newFixedThreadPool(count);
+    int r = 0;
+    while (r < 50) {
```

Comment:
It's still not completely clear for me but you've done your homework. Please make those named constants and if possible add some intuition why they are what they are. Some of those magic constants my be driven by the Google implementation which may change in the future.